photon-api/src/main/scala/com/linkedin/photon/ml/data/LocalDataset.scala

```diff
@@ -180,11 +180,91 @@ object LocalDataset {
   }
 
   /**
-   * Compute Pearson correlation scores.
+   * Compute Pearson correlation scores using a numerically stable algorithm.
    *
    * @param labelAndFeatures An array of (label, feature) tuples
    * @return The Pearson correlation scores for each tuple
    */
+  protected[ml] def stableComputePearsonCorrelationScore(
+      labelAndFeatures: Array[_ <: (Double, Vector[Double])]): Map[Int, Double] = {
+
+    val featureMeans = mutable.Map[Int, Double]()
+    var featureUnscaledVars = mutable.Map[Int, Double]()
+    var labelMean = 0.0
+    var labelUnscaledVariance = 0.0
+    val unscaledCovariances = mutable.Map[Int, Double]()
+    var interceptAdded = false
+    var numSamples = 0
+
+    labelAndFeatures.foreach { case (label, features) =>
+      numSamples += 1
+
+      val deltaLabel = label - labelMean
+      labelMean += deltaLabel / numSamples
+      labelUnscaledVariance += deltaLabel * (label - labelMean)
+
+      // Note that, if there is duplicated keys in the feature vector, then the following Pearson correlation scores
+      // calculation will screw up
+      features.iterator.foreach { case (key, value) =>
+        val prevFeatureMean = featureMeans.getOrElse(key, 0.0)
+        val deltaFeature = value - prevFeatureMean
+        val featureMean = prevFeatureMean + deltaFeature / numSamples
+
+        val prevFeatureUnscaledVar = featureUnscaledVars.getOrElse(key, 0.0)
+        val featureUnscaledVar = prevFeatureUnscaledVar + deltaFeature * (value - featureMean)
+
+        val prevCovariance = unscaledCovariances.getOrElse(key, 0.0)
+        val unscaledCovariance = prevCovariance + deltaFeature * deltaLabel * (numSamples - 1) / numSamples
+
+        featureMeans.update(key, featureMean)
+        featureUnscaledVars.update(key, featureUnscaledVar)
+        unscaledCovariances.update(key, unscaledCovariance)
+      }
+    }
+
+    val labelStd = math.sqrt(labelUnscaledVariance)
+
+    featureMeans
+      .iterator
+      .map { case (key, featureMean) =>
+        val featureStd = math.sqrt(featureUnscaledVars(key))
+        val covariance = unscaledCovariances(key)
+
+        // When the standard deviation of the feature is close to 0 we treat it as the intercept term.
+        val score = if (featureStd < math.sqrt(numSamples) * MathConst.EPSILON) {
+          // Note that if the mean and standard deviation are equal to zero, it either means that the feature is constant
+          if (interceptAdded || featureMean == 0.0) {
+            0.0
+          } else {
+            interceptAdded = true
+            1.0
```

Comment:
Should the score always be 0.0 when we compute the correlation between the label and any constant?

Reply:
The correlation between a constant and a random variable is ill-defined. One could argue that it is 0 [1], but I think here it's more relevant to keep in mind hoe the function will be used. 

Since it will be used to filter-out features, we do want to get rid of all-but-one constant features, this allows to get rid of the co-linearity of the many constant columns.

So since  I would argue that we want to keep only one "intercept" column, we would want to give a low Pearson Correlation score in order to get rid of these correlations first.

[1] https://stats.stackexchange.com/a/267155